wagtail/fields.py

```diff
@@ -202,15 +202,11 @@ def get_prep_value(self, value):
             )
         else:
             # When querying with JSONField features, the rhs might not be a StreamValue.
-            # Note: when Django 4.2 is the minimum supported version, this can be removed
-            # as the serialisation is handled in get_db_prep_value instead.
             return self.json_field.get_prep_value(value)
 
     def get_db_prep_value(self, value, connection, prepared=False):
         if not isinstance(value, StreamValue):
             # When querying with JSONField features, the rhs might not be a StreamValue.
-            # As of Django 4.2, JSONField value serialisation is handled in
-            # get_db_prep_value instead of get_prep_value.
             return self.json_field.get_db_prep_value(value, connection, prepared)
         return super().get_db_prep_value(value, connection, prepared)
```

Comment:
Sorry, my comment in https://github.com/wagtail/wagtail/pull/9693#discussion_r1027830773 was indeed inaccurate.

> When we only support Django 4.2 and up, this can be removed and we can revert to the original logic pre-`use_json_field` by replacing the `elif` above with `else`:

"the original logic pre-`use_json_field`" did not have any handling for JSONField querying features (because it wasn't a thing back then 🙂), which is what the `else` clause covers. So we _can't_ remove the `else` clause if we want to keep the support for JSONField lookups and transforms.

If I understood correctly, before adding `use_json_field` (and support for JSONField lookups and transforms), it was pretty much guaranteed/expected that the `value` when it falls after the first `if` clause is a `StreamValue` – or at least, if it's non-empty. Hence, the code was:

https://github.com/wagtail/wagtail/blob/cbe9dc729b7cc92229b535c9bedd719b1a12c9e8/wagtail/core/fields.py#L137-L145

Where `BaseStreamBlock.get_prep_value()` can confidently call `value.get_prep_value()` without checking whether `value` really is a `StreamValue`:

https://github.com/wagtail/wagtail/blob/a7ac18f6091371c4d287f3b49b9932c63ba4267f/wagtail/blocks/stream_block.py#L277-L286

When we added `use_json_field` and the lookups and transforms support, there is a higher probability where `get_prep_value()` is called with a non-`StreamValue`. Hence, I added the `elif`:

https://github.com/wagtail/wagtail/blob/dcae64c255f2fe97f658b1a3f438d3644b197661/wagtail/fields.py#L190-L198

to essentially say: if we're still using the old text-based StreamField, we can go ahead with the old code that delegates the logic to `BaseStreamBlock.get_prep_value()` – even if `value` is not a `StreamValue` somehow, which wasn't considered at all before `use_json_field` was added. But otherwise (i.e. JSONField is used and the value is not `StreamValue`), let JSONField do its thing.

So, instead of replacing the `elif` with `else`, the more accurate interpretation of my comment's intention is something like:

```diff
diff --git a/wagtail/fields.py b/wagtail/fields.py
index 7d2f60bb40..2e53ca5dc2 100644
--- a/wagtail/fields.py
+++ b/wagtail/fields.py
@@ -201,17 +201,24 @@ class StreamField(models.Field):
                 self.stream_block.get_prep_value(value), cls=DjangoJSONEncoder
             )
         else:
-            # When querying with JSONField features, the rhs might not be a StreamValue.
-            # Note: when Django 4.2 is the minimum supported version, this can be removed
-            # as the serialisation is handled in get_db_prep_value instead.
+            # If the value is not a StreamValue, it's likely the field is being
+            # used in a non-Wagtail context, e.g. in queries with JSONField features.
+            # For such cases, we return the value without serializing it, and let
+            # Django handle it with its own way instead. We can either do
+            # self.json_field.get_prep_value(value) here, or use
+            # super().get_prep_value(value) – they're the same thing because
+            # JSONField no longer overrides get_prep_value, and the serialization
+            # process is done in get_db_prep_value instead.
             return self.json_field.get_prep_value(value)

     def get_db_prep_value(self, value, connection, prepared=False):
         if not isinstance(value, StreamValue):
             # When querying with JSONField features, the rhs might not be a StreamValue.
-            # As of Django 4.2, JSONField value serialisation is handled in
-            # get_db_prep_value instead of get_prep_value.
+            # Delegate the rest of the get_db_prep_value process to the JSONField,
+            # which includes the serialization process.
             return self.json_field.get_db_prep_value(value, connection, prepared)
+        # For StreamValues, we've always been handling the serialization in get_prep_value,
+        # so we just use the base Field implementation of get_db_prep_value here.
         return super().get_db_prep_value(value, connection, prepared)

     def from_db_value(self, value, expression, connection):
```

So, from what I understand, we have different options here...

1. Keep using `self.json_field.get_prep_value()`, even though it's exactly the same as `super().get_prep_value()`. We may want to do this in case Django ends up overriding `get_prep_value()` again in the future.
2. Use `super().get_prep_value()` instead. [This is basically `return value` as-is, but it also handles lazy objects.](https://github.com/django/django/blob/f8ef4579ea710f93ec7edc93c6f3f216bd55d6be/django/db/models/fields/__init__.py#L988-L992)
3. Follow Django's move by implementing the serialisation in `get_db_prep_value()` instead of `get_prep_value()`. This may look like the following:
    ```diff
    diff --git a/wagtail/fields.py b/wagtail/fields.py
    index 7d2f60bb40..79eae7a66e 100644
    --- a/wagtail/fields.py
    +++ b/wagtail/fields.py
    @@ -3,7 +3,6 @@ import json
    from django.core.serializers.json import DjangoJSONEncoder
    from django.core.validators import MaxLengthValidator
    from django.db import models
    -from django.db.models.fields.json import KeyTransform
    from django.utils.encoding import force_str
    from django.utils.functional import cached_property

    @@ -185,6 +184,7 @@ class StreamField(models.Field):
            return result

        def get_prep_value(self, value):
    +        value = super().get_prep_value(value)
            if (
                isinstance(value, StreamValue)
                and not (value)
    @@ -197,36 +197,37 @@ class StreamField(models.Field):
                return value.raw_text
            elif isinstance(value, StreamValue):
                # StreamValue instances must be prepared first.
    -            return json.dumps(
    -                self.stream_block.get_prep_value(value), cls=DjangoJSONEncoder
    -            )
    +            return self.stream_block.get_prep_value(value)
            else:
    -            # When querying with JSONField features, the rhs might not be a StreamValue.
    -            # Note: when Django 4.2 is the minimum supported version, this can be removed
    -            # as the serialisation is handled in get_db_prep_value instead.
    -            return self.json_field.get_prep_value(value)
    +            # If the value is not a StreamValue, it's likely the field is being
    +            # used in a non-Wagtail context, e.g. in queries with JSONField features.
    +            return super().get_prep_value(value)

        def get_db_prep_value(self, value, connection, prepared=False):
    -        if not isinstance(value, StreamValue):
    -            # When querying with JSONField features, the rhs might not be a StreamValue.
    -            # As of Django 4.2, JSONField value serialisation is handled in
    -            # get_db_prep_value instead of get_prep_value.
    -            return self.json_field.get_db_prep_value(value, connection, prepared)
    -        return super().get_db_prep_value(value, connection, prepared)
    +        # Use JSONField's get_db_prep_value method to handle the serialization,
    +        # which may differ between database backends. However, use our own
    +        # get_prep_value method to ensure that StreamValue instances are prepared
    +        # before being passed to JSONField.
    +        if not prepared:
    +            value = self.get_prep_value(value)
    +        return self.json_field.get_db_prep_value(
    +            value, connection=connection, prepared=True
    +        )

        def from_db_value(self, value, expression, connection):
    -        if isinstance(expression, KeyTransform):
    -            # This could happen when using JSONField key transforms,
    -            # e.g. Page.object.values('body__0').
    -            try:
    -                # We might be able to properly resolve to the appropriate StreamValue
    -                # based on `expression` and `self.stream_block`, but it might be too
    -                # complicated to do so. For now, just deserialise the value.
    -                return json.loads(value)
    -            except ValueError:
    -                # Just in case the extracted value is not valid JSON.
    -                return value
    -
    +        # Historically, StreamField's deserialization used to be handled by
    +        # to_python, which in turn handled by BaseStreamBlock.to_python. This was
    +        # always the case even before and after the use of the JSON data type.
    +
    +        # However, now that we can be confident all StreamField data has been
    +        # migrated to use JSON in the database, we can reuse any special handling
    +        # that JSONField.from_db_value provides, e.g. for handling KeyTransforms
    +        # on SQLite.
    +
    +        # This means we are passing a deserialized value to StreamBlock.to_python,
    +        # which is a change from the previous behaviour. However, this is fine
    +        # because to_python can handle both serialized and deserialized values.
    +        value = self.json_field.from_db_value(value, expression, connection)
            return self.to_python(value)

        def formfield(self, **kwargs):
    @@ -239,8 +240,23 @@ class StreamField(models.Field):
            return super().formfield(**defaults)

        def value_to_string(self, obj):
    +        # This method is used for serialization using django.core.serializers,
    +        # which is used by dumpdata and loaddata for serializing model objects.
    +        # Unlike other fields, JSONField only uses value_from_object without
    +        # doing the actual serialization, so that it doesn't end up being
    +        # double-serialized when the model object is serialized.
    +
    +        # Unfortunately, this is also used by django-modelcluster, which is used
    +        # to serialize model objects to be stored in revisions. When we migrated
    +        # StreamField to use the JSON data type, we did not change this method's
    +        # behaviour, i.e. it still returns a JSON-shaped string, to ensure that
    +        # revisions are still saved in the same format as before – even if it
    +        # means StreamField inside the revision data becomes double-serialized.
    +
    +        # Now that we change get_prep_value to not do the serialization in favor
    +        # of get_db_prep_value, we need to add the serialization here too.
            value = self.value_from_object(obj)
    -        return self.get_prep_value(value)
    +        return json.dumps(self.get_prep_value(value), cls=self.json_field.encoder)

        def get_searchable_content(self, value):
            return self.stream_block.get_searchable_content(value)
    ```
   I think this is the best middle ground between keeping most of our StreamField implementation and reusing Django's JSONField code where we can. However, I also understand that this might be a bit controversial due to the amount of changes to StreamField's inner workings.
4. Something more drastic, refactor `StreamField` to be a proper subclass of `JSONField`. Likely needs refactoring in a lot of other places too. Not worth it imo.

Reply:
Oh, I almost forgot! An alternative to option (1) and (2) is: since Django no longer skips `get_prep_value()`, we can put the serialization in `get_prep_value()` directly and remove the need to override `get_db_prep_value()`, similar to the logic before Django moved the serialization to `get_db_prep_value()`:

```diff
diff --git a/wagtail/fields.py b/wagtail/fields.py
index 7d2f60bb40..88728b1065 100644
--- a/wagtail/fields.py
+++ b/wagtail/fields.py
@@ -201,18 +201,17 @@ class StreamField(models.Field):
                 self.stream_block.get_prep_value(value), cls=DjangoJSONEncoder
             )
         else:
-            # When querying with JSONField features, the rhs might not be a StreamValue.
-            # Note: when Django 4.2 is the minimum supported version, this can be removed
-            # as the serialisation is handled in get_db_prep_value instead.
-            return self.json_field.get_prep_value(value)
-
-    def get_db_prep_value(self, value, connection, prepared=False):
-        if not isinstance(value, StreamValue):
-            # When querying with JSONField features, the rhs might not be a StreamValue.
-            # As of Django 4.2, JSONField value serialisation is handled in
-            # get_db_prep_value instead of get_prep_value.
-            return self.json_field.get_db_prep_value(value, connection, prepared)
-        return super().get_db_prep_value(value, connection, prepared)
+            # If the value is not a StreamValue, it's likely the field is being
+            # used in a non-Wagtail context, e.g. in queries with JSONField features.
+            # For such cases, we only need to serialise the value.
+            # Ideally, we should make use of JSONField's serialization logic, but
+            # it has since been moved to `get_db_prep_value`, which is called
+            # before this method. However, we don't make use of JSONField's
+            # db-specific serialization in the above StreamValue cases anyway,
+            # so we might as well just serialise the value here.
+            return json.dumps(
+                self.json_field.get_prep_value(value), cls=self.json_field.encoder
+            )

     def from_db_value(self, value, expression, connection):
         if isinstance(expression, KeyTransform):
```

However, we'd lose a few of the special handling that JSONField has, such as for handling `Value(value, output_field=JSONField())` and handling compile-able expressions:

https://github.com/django/django/blob/f8ef4579ea710f93ec7edc93c6f3f216bd55d6be/django/db/models/fields/json.py#L99-L107